target/riscv/op_helper_cheri.c

```diff
@@ -160,12 +177,16 @@ void HELPER(cspecialrw)(CPUArchState *env, uint32_t cd, uint32_t cs,
 
     assert(index <= 31 && "Bug in translator?");
     enum SCRAccessMode mode = scr_info[index].access;
-    if (mode == SCR_Invalid || (cs != 0 && !scr_info[index].w)) {
+    bool is_write = cs != 0;
+    if (mode == SCR_Invalid || (is_write && !scr_info[index].w)) {
         riscv_raise_exception(env, RISCV_EXCP_ILLEGAL_INST,
                               _host_return_address);
     }
     bool can_access_sysregs = cheri_have_access_sysregs(env);
-    if (scr_needs_asr(mode) && !can_access_sysregs) {
+    if (is_write && scr_needs_asr_w(mode) && !can_access_sysregs) {
+        raise_cheri_exception(env, CapEx_AccessSystemRegsViolation, 32 + index);
+    }
+    if (scr_needs_asr_r(mode) && !can_access_sysregs) {
```

Comment:
Doesn't currently matter since it will pass in all current cases but I think we also need a is_read check here